employee-rostering-gwtui/src/main/java/org/optaweb/employeerostering/gwtui/client/gwtjackson/ServerSideExceptionDeserializer.java

```diff
@@ -0,0 +1,110 @@
+/*
+ * Copyright 2019 Red Hat, Inc. and/or its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.optaweb.employeerostering.gwtui.client.gwtjackson;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+
+import javax.inject.Singleton;
+
+import com.github.nmorel.gwtjackson.client.stream.JsonReader;
+import com.github.nmorel.gwtjackson.client.stream.JsonToken;
+import com.github.nmorel.gwtjackson.client.stream.impl.DefaultJsonReader;
+import com.github.nmorel.gwtjackson.client.stream.impl.StringReader;
+import org.optaweb.employeerostering.shared.exception.ServerSideException;
+
+@Singleton
+public class ServerSideExceptionDeserializer {
+
+    // Work around for https://github.com/nmorel/gwt-jackson/issues/79
+    // (Cannot use the gwt-jackson ObjectMapper due to a compilation error)
+    public ServerSideException deserializeFromJsonString(String json) {
+        JsonReader reader = new DefaultJsonReader(new StringReader(json));
+        try {
+            return deserialize(reader, json);
+        } finally {
+            reader.close();
+        }
+    }
+
+    public ServerSideException deserialize(JsonReader reader, String json) {
+        Set<String> fields = new HashSet<>(Arrays.asList("i18nKey", "messageParameters", "exceptionClass", "stackTrace", "exceptionMessage", "exceptionCause"));
```

Comment:
Do I get it correctly that this is a list of all fields of the `ServerSideException` class? If so, it's very fragile - there is no way to make sure every time you add a field to the `ServerSideException` you add a corresponding piece of code here.

Reply:
It is suppose to be fragile; having it like this make it fail fast; alerting me that I need to update this piece of code. We don't have reflection on the client, and we need to make sure this code navigate every field of the ServerSideException class, so we have few options. Technically, this class shouldn't be needed, but I am getting a compile error in GWT Devmode (https://github.com/nmorel/gwt-jackson/issues/79). A workaround is to add "dagger-compiler" as a dependency, but I rather not add otherwise unneeded dependencies.

Reply:
I've seen the issue and agree with you that bringing in another dependency as a workaround is far from ideal solution. Reflection is slightly better, still problematic solution in my opinion. But if you say it cannot be compiled to the client code, we don't have to even talk about this option.

Let's create the enum and reference it from the `ServerSideException` to minimize the maintenance cost. And wherever you need to work with the field, use the enum please. Are you ok with that?